cluster_experiments/inference/analysis_plan.py

```diff
@@ -0,0 +1,297 @@
+    variants : List[Variant]
+        A list of Variant instances
+    variant_col : str
+        name of the column with the experiment groups
+    alpha : float
+        significance level used to construct confidence intervals
+    """
+
+    def __init__(
+        self,
+        tests: List[HypothesisTest],
+        variants: List[Variant],
+        variant_col: str = "treatment",
+        alpha: float = 0.05,
+    ):
+        """
+        Parameters
+        ----------
+        tests : List[HypothesisTest]
+            A list of HypothesisTest instances
+        variants : List[Variant]
+            A list of Variant instances
+        variant_col : str
+            The name of the column containing the variant names.
+        alpha : float
+            significance level used to construct confidence intervals
+        """
+
+        self.tests = tests
+        self.variants = variants
+        self.variant_col = variant_col
+        self.alpha = alpha
+
+        self._validate_inputs()
+
+    def _validate_inputs(self):
+        """
+        Validates the inputs for the AnalysisPlan class.
+
+        Raises
+        ------
+        TypeError
+            If tests is not a list of HypothesisTest instances or if variants is not a list of Variant instances.
+        ValueError
+            If tests or variants are empty lists.
+        """
+        if not isinstance(self.tests, list) or not all(
+            isinstance(test, HypothesisTest) for test in self.tests
+        ):
+            raise TypeError("Tests must be a list of HypothesisTest instances")
+        if not isinstance(self.variants, list) or not all(
+            isinstance(variant, Variant) for variant in self.variants
+        ):
+            raise TypeError("Variants must be a list of Variant instances")
+        if not isinstance(self.variant_col, str):
+            raise TypeError("Variant_col must be a string")
+        if not self.tests:
+            raise ValueError("Tests list cannot be empty")
+        if not self.variants:
+            raise ValueError("Variants list cannot be empty")
+
+    def analyze(
+        self,
+        exp_data: pd.DataFrame,
+        pre_exp_data: Optional[pd.DataFrame] = None,
+        verbose: bool = False,
+    ) -> AnalysisPlanResults:
+        """
+        Method to run the experiment analysis.
+        """
+
+        # Validate input data at the beginning
+        self._validate_data(exp_data, pre_exp_data)
+
+        analysis_results = EmptyAnalysisPlanResults()
+
+        for test in self.tests:
+            if test.is_cupac:
+                exp_data = test.cupac_handler.add_covariates(
+                    df=exp_data, pre_experiment_df=pre_exp_data
+                )
+
+            target_col = test.metric.get_target_column_from_metric()
+
+            for treatment_variant in self.treatment_variants:
+                for dimension in test.dimensions:
+                    for dimension_value in dimension.iterate_dimension_values():
```

Comment:
I would prefer to keep it as is, because the test knows what a dimension is, given the fact that it takes a list of dimensions as attributes when instantiating a test object. This way we use a prepare_data() method at the test level

Reply:
ok I see your point below about self.dimensions not being utilised in hypotesis_test directly. I'll leave this as an open topic as I'm not sure of a better solution

Reply:
let's see if this makes sense: we change analysisplan to have this:

```python

    def __init__(
        self,
        tests: List[HypothesisTest],
        variants: List[Variant],
        dimensions: List[Dimension],
        variant_col: str = "treatment",
        alpha: float = 0.05,
    ):
```

and the filter is done on the dimension,

changing this loop for

```
            for treatment_variant in self.treatment_variants:
                for dimension in self.dimensions:
                    for dimension_value in dimension.iterate_dimension_values():
```

Two questions:
* Does this make sense or am I missing something?
* Would this change the main interfaces (if it doesn't, we could leave it for later)?

Reply:
this is a viable solution but it has the limitation that you force the user to compute results splitting by the same dimensions all the metrics in the same analysis plan. Theoretically the user could create different analysis plans for different dimensions even if the data frame is the same, in that case it would have the same flexibility at the cost of creating extra analysis plans.

Wdyt about this?

Reply:
No strong opinion, your call

Reply:
ok, let's leave it as it is right now in favour of increased flexibility! I'll proceed with the tests